src/blockdev.rs

```diff
@@ -105,37 +146,72 @@ impl Disk {
+                path: devinfo.name.to_owned(),
+                label: devinfo.label.clone(),
+                fstype: devinfo.fstype.clone(),
+                parent: self.path.to_owned(),
+                mountpoint,
+                swap,
+            });
+        }
+        Ok(result)
+    }
+
     fn get_partitions(&self, with_holders: bool) -> Result<Vec<Partition>> {
         // walk each device in the output
         let mut result: Vec<Partition> = Vec::new();
-        for devinfo in lsblk(Path::new(&self.path), true)? {
-            if let Some(name) = devinfo.get("NAME") {
-                match devinfo.get("TYPE") {
-                    // If unknown type, skip.
-                    None => continue,
-                    // If whole-disk device, skip.
-                    Some(t) if t == &"disk".to_string() => continue,
-                    // If partition, allow.
-                    Some(t) if t == &"part".to_string() => (),
-                    // If with_holders is true, allow anything else.
-                    Some(_) if with_holders => (),
-                    // Ignore LVM or RAID devices which are using one of the
-                    // partitions but aren't a partition themselves.
-                    _ => continue,
-                };
-                let (mountpoint, swap) = match devinfo.get("MOUNTPOINT") {
-                    Some(mp) if mp == "[SWAP]" => (None, true),
-                    Some(mp) => (Some(mp.to_string()), false),
-                    None => (None, false),
-                };
-                result.push(Partition {
-                    path: name.to_owned(),
-                    label: devinfo.get("LABEL").map(<_>::to_string),
-                    fstype: devinfo.get("FSTYPE").map(<_>::to_string),
-                    parent: self.path.to_owned(),
-                    mountpoint,
-                    swap,
-                });
+        let deviceinfo = Device::lsblk(Path::new(&self.path), true)?;
+        let mut partition = self.compute_partition(&deviceinfo, with_holders)?;
+        if !partition.is_empty() {
+            result.append(&mut partition);
+        }
+        let mut childpartitioninfo = self.get_partitions_recurse(&deviceinfo, with_holders)?;
+        if !childpartitioninfo.is_empty() {
+            result.append(&mut childpartitioninfo);
+        }
+        Ok(result)
+    }
+
+    fn get_partitions_recurse(
+        &self,
+        deviceinfo: &Device,
+        with_holders: bool,
+    ) -> Result<Vec<Partition>> {
+        let mut result: Vec<Partition> = Vec::new();
+        if let Some(children) = deviceinfo.children.as_ref() {
+            if !children.is_empty() {
+                for child in children {
+                    let mut childpartitions = self.compute_partition(child, with_holders)?;
+                    result.append(&mut childpartitions);
+                    self.get_partitions_recurse(child, with_holders)?;
+                }
```

Comment:
OK digging more into this took me to https://github.com/coreos/coreos-installer/pull/648. Let's do that first, and then this should essentially be just `get_partitions()` enumerating over the immediate children only for partitions (so no need for recursion).

Also, maybe as a follow-up we should investigate just returning `Vec<Device>` here instead of a separate `Partition` type.